### PR TITLE
Add ability to run Postgres proxy on separate listener

### PR DESCRIPTION
**Pull request by smallinsky**
# What 
Add ability to run Postgres proxy on separate listener https://github.com/gravitational/teleport/issues/6828

# How: 
Create separate Postgres listener instead of multiplexing Postgres service on Web port multiplexer if  `postgres_listen_addr` setting was explicitly provided. 
A new `postgres_listen_addr` address is propagated to the tsh clients in PingResponse message by a new filed `DBProxySettings.PostgresListenerAddr` but for legacy tsh clients the `PostgresPublicAddr` field was used if `postgres_public_addr` was not set to support old tsh clients where the `DBProxySettings.PostgresListenerAddr` field in is not yet available.

### CODE REVIEW

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:2553]: You're already passing `cfg` into this function so I would get rid of the `useSeparatePostgresListener` parameter and just do `!cfg.Proxy.PostgresAddr.IsEmpty()` inside the function again. Would be a bit cleaner interface.

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:2621]:
```suggestion
// setPostgresListener start Postgres proxy listener based on configuration settings. By default, Postgres service is
// multiplexed on Teleport Proxy web port but if the postgres_listen_addr flag was provided, the
// Postgres service runs on a separate port taken from it.
```

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:3317]:
```suggestion
		proxySettings.DB.PostgresListenAddr = cfg.Proxy.PostgresAddr.String()
```

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:3335]:
```suggestion
// getPostgresPublicAddr returns the proxy PostgresPublicAddrs based on whether the Postgres proxy service
```

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:3343]: Is this condition correct? It seems like `len(cfg.Proxy.PostgresPublicAddrs) != 0` would already be handled by the `> 0` check above so it would never be true here?

File: lib/config/configuration.go
Comment (r0mant) [reply lib/config/configuration.go:811]: Also, why is this checking `fc` (file config) but returning the value from `cfg` (service config)?

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:3357]: Is there a way to reuse the port calculation logic from `getPostgresDefaultPort` you have in `lib/config/configuration.go` above? Seems like it does basically the same thing. And that function already operates on `service.Config` which you have here.

So I would move that function here (I don't think `fc` is really needed there - I left a comment there) and use both here and in lib/config as well so we don't end up with duplication.

Maybe even make a helper method on the `service.Config` struct.

**Review by Joerger — APPROVED** [review]
LGTM once you've resolved remaining comments

File: lib/service/service.go
Comment (smallinsky) [reply lib/service/service.go:3343]: fixed

File: lib/config/configuration.go
Comment (smallinsky) [reply lib/config/configuration.go:811]: > Also, why is this checking fc (file config) but returning the value from cfg (service config)?

Fixed. 

> Shouldn't the port from Postgres public address (if it's specified) take precedence?

In case of proxy configuration no, the `PostgresAddr` is a local proxy listener address and in case of local proxy listener the PublicAddress port should not take precedence over local proxy listener port. 

though `PostgresPublicAddr` should take precedence over local port in case of Proxy Ping response but for this a separate function is used.

File: lib/service/service.go
Comment (smallinsky) [reply lib/service/service.go:3357]: > Is there a way to reuse the port calculation logic from getPostgresDefaultPort you have in lib/config/configuration.go above?

The`getPostgresDefaultPort` function returns the postgres port. In this function if the PostgresPublicAddrs was not set  the port is constant - taken from `cfg.Proxy.PostgresAddr.Port` thus  I don't see how this logic can be combined.

File: lib/service/proxy_settings.go
Comment (r0mant) [inline lib/service/proxy_settings.go:123]: Since you added this, is the block above on lines 117-119 still needed? Seems like it can be removed now:

```go
	if len(p.cfg.Proxy.PostgresPublicAddrs) > 0 {
		settings.DB.PostgresPublicAddr = p.cfg.Proxy.PostgresPublicAddrs[0].String()
	}
```

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:2633]: Is this going to potentially mess up the listener that `process.setPostgresListener` set right above?

File: lib/srv/db/proxyserver.go
Comment (r0mant) [inline lib/srv/db/proxyserver.go:257]: Seems like this function handles only Postgres connections (separate listener or not) anyway, is there any point in still having it? Let's get rid of it and just use `s.PostgresProxy()` directly in where all Postgres connections are handled?

**Review by r0mant — APPROVED** [review]
lgtm overall but please take a look at a couple questions/suggestions before merging.

**Comment by russjones** [comment]
@smallinsky @r0mant Can we merge this PR?

**Pull request merged**
